internal/addonsdk/sdk.go

```diff
@@ -0,0 +1,215 @@
+	statusReporterSingletonMutex = &sync.Mutex{}
+)
+
+// a nice heartbeat-reporter implemented by the MT-SRE which our tenants can use
+// if they don't like it, they can implement their own heartbeat reporter by creating a type which implements the `addonsdk.statusReporterClient` interface
+type StatusReporter struct {
+	// object provided by the client/tenants which implements the addonsdk.client interface
+	addonInstanceInteractor client
+	addonName               string
+	addonTargetNamespace    string
+
+	// the latest conditions which the heartbeat reporter would be reporting periodically
+	latestConditions []metav1.Condition
+
+	// to control the rate at which the heartbeat reporter would run
+	interval time.Duration
+	ticker   *time.Ticker
+
+	// for effectively communicating the stop and update signals
+	stopperCh chan bool
+	updateCh  chan updateOptions
+
+	// for concurrency-safely executing one instance of heartbeat reporter loop
+	executeOnce sync.Once
+
+	//for tracking if the heartbeat reporter is running or done running
+	doneCh chan bool
+
+	log logr.Logger
+}
+
+// ensure that the `StatusReporter` implements the `addonsdk.statusReporterClient` interface
+var _ statusReporterClient = (*StatusReporter)(nil)
+
+// InitializeStatusReporterSingleton sets up a singleton of the type `StatusReporter` (only if it doesn't exist yet) and returns it to the caller.
+func InitializeStatusReporterSingleton(addonInstanceInteractor client, addonName string, addonTargetNamespace string, logger logr.Logger) *StatusReporter {
+	if statusReporterSingleton == nil {
+		statusReporterSingletonMutex.Lock()
+		defer statusReporterSingletonMutex.Unlock()
+		if statusReporterSingleton == nil {
+			statusReporterSingleton = &StatusReporter{
+				addonInstanceInteractor: addonInstanceInteractor,
+				addonName:               addonName,
+				addonTargetNamespace:    addonTargetNamespace,
+				latestConditions: []metav1.Condition{
+					{
+						Type:    "addons.managed.openshift.io/Healthy",
+						Status:  "False",
+						Reason:  "NoHeartbeatReported",
+						Message: fmt.Sprintf("Addon '%s' hasn't reported any heartbeat yet", addonName),
+					},
+				},
+				stopperCh: make(chan bool),
+				updateCh:  make(chan updateOptions),
+				doneCh:    make(chan bool),
+				log:       logger,
+			}
+			// because the heartbeat reporter still hasn't been started
+			defer close(statusReporterSingleton.doneCh)
+		}
+	}
+
+	return statusReporterSingleton
+}
+
+func GetStatusReporterSingleton() (*StatusReporter, error) {
+	if statusReporterSingleton == nil {
+		return nil, fmt.Errorf("heartbeat-reporter not found to be initialised. Initialize it by calling `InitializeStatusReporterSingleton(...)`")
+	}
+	return statusReporterSingleton, nil
+}
+
+//nolint
+func (sr StatusReporter) GetAddonTargetNamespace() string {
+	return sr.addonTargetNamespace
+}
+
+//nolint
+func (sr StatusReporter) LatestConditions() []metav1.Condition {
+	return sr.latestConditions
+}
```

Comment:
> Lint was getting triggered as it noticed that StatusReporter has a mutex-compliant attribute (executeOnce sync.Once) due to which it was insisting on having pointer receivers.
And the linter is very right to complain in this case, because you are making a copy of the mutex lock and you have disabled the linter to tell people about it.
Anyone using a mutex lock within this method, will be surprised to see it not working...

`LatestConditions` also has a data-race, because the variable is concurrently accessed by the worker goroutine.

In general we should _never*_ disable linters, because they complain for a reason.
If we find ourselves in a situation where a linter is reporting a false positive, we want to only disable the offending linter and leave a comment:
https://github.com/openshift/addon-operator/blob/eaa47862db365decf8d9ff36814584de3fb3c74c/integration/helpers.go#L308

> but I don't see a "client-concerning" data-racey situation with calling `LatestConditions()`

There is no save data race. If you share memory, you have to protect it.

> Let me know in case you still want the GetAddonTargetNamespace() to have pointer receiver for some reason which I missed.

I have 3 issues with this function:
1. The disabled linter
2. `In general, all methods on a given type should have either value or pointer receivers, but not a mixture of both.`
https://go.dev/tour/methods/8
3. why do we even need it?